File: src/facility_check.cpp
#include <spdlog/spdlog.h>
#include <filesystem>
#include <fstream>
#include <sstream>

#include "dcpdoctor/facility_check.h"
#include "dcpdoctor/checksum_verify.h"
#include "dcpdoctor/compliance.h"
#include "dcpdoctor/cpl.h"
#include "dcpdoctor/isdcf.h"
#include "dcpdoctor/validators.h"

namespace fs = std::filesystem;

namespace dcpdoctor
{

static CheckItem make_item(const std::string& category, const std::string& name, bool passed,
                           const std::string& detail = {}, const std::string& severity = "error")
{
  CheckItem item;
  item.category = category;
  item.check_name = name;
  item.passed = passed;
  item.detail = detail;
  item.severity = severity;
  return item;
}

FacilityCheckResult run_facility_check(const FacilityCheckOptions& opts)
{
  FacilityCheckResult result;

  if(!fs::exists(opts.dcp_dir) || !fs::is_directory(opts.dcp_dir))
  {
    result.error = "DCP directory not found: " + opts.dcp_dir.string();
    return result;
  }

  // --- Structure checks ---

  // Check ASSETMAP exists
  bool has_assetmap =
      fs::exists(opts.dcp_dir / "ASSETMAP") || fs::exists(opts.dcp_dir / "ASSETMAP.xml");
  result.items.push_back(make_item("structure", "ASSETMAP present", has_assetmap,
                                   has_assetmap ? "" : "Missing ASSETMAP or ASSETMAP.xml"));

  // Check VOLINDEX exists
  bool has_volindex =
      fs::exists(opts.dcp_dir / "VOLINDEX") || fs::exists(opts.dcp_dir / "VOLINDEX.xml");
  result.items.push_back(make_item("structure", "VOLINDEX present", has_volindex,
                                   has_volindex ? "" : "Missing VOLINDEX or VOLINDEX.xml"));

  // Find PKL files
  std::vector<fs::path> pkls;
  for(const auto& entry : fs::directory_iterator(opts.dcp_dir))
  {
    if(!entry.is_regular_file())
      continue;
    auto ext = entry.path().extension().string();
    if(ext == ".xml")
    {
      std::ifstream f(entry.path());
      std::string line;
      for(int i = 0; i < 5 && std::getline(f, line); ++i)
      {
        if(line.find("PackingList") != std::string::npos)
        {
          pkls.push_back(entry.path());
          break;
        }
      }
    }
  }
  result.items.push_back(make_item("structure", "PKL present", !pkls.empty(),
                                   pkls.empty() ? "No PackingList XML found" : ""));

  // Find CPL files
  std::vector<fs::path> cpls;
  for(const auto& entry : fs::directory_iterator(opts.dcp_dir))
  {
    if(!entry.is_regular_file())
      continue;
    auto cpl = Cpl::parse(entry.path());
    if(cpl.has_value() && !cpl->id.empty())
      cpls.push_back(entry.path());
  }
  result.items.push_back(make_item("structure", "CPL present", !cpls.empty(),
                                   cpls.empty() ? "No CompositionPlaylist XML found" : ""));

  // Check CPL has reels
  for(const auto& cpl_path : cpls)
  {
    auto cpl = Cpl::parse(cpl_path);
    if(cpl.has_value())
    {
      result.items.push_back(make_item("structure", "CPL has reels", !cpl->reels.empty(),
                                       cpl->reels.empty() ? "CPL has no reels/segments" : ""));
    }
  }

  // Check MXF files exist
  bool has_mxf = false;
  for(const auto& entry : fs::directory_iterator(opts.dcp_dir))
  {
    if(entry.path().extension() == ".mxf")
    {
      has_mxf = true;
      break;
    }
  }
  result.items.push_back(make_item("structure", "MXF essence files present", has_mxf,
                                   has_mxf ? "" : "No .mxf files found in DCP directory"));

  // --- Compliance checks ---
  auto compliance_notes = check_smpte_compliance(opts.dcp_dir, opts.expected_standard, opts.strict);
  bool compliance_ok = true;
  for(const auto& note : compliance_notes)
  {
    if(note.severity == Severity::error)
    {
      compliance_ok = false;
      break;
    }
  }
  result.items.push_back(make_item(
      "compliance", "SMPTE compliance", compliance_ok,
      compliance_ok ? "" : std::to_string(compliance_notes.size()) + " compliance issue(s)"));

  // Add individual compliance notes as items
  for(const auto& note : compliance_notes)
  {
    std::string sev = "info";
    if(note.severity == Severity::error)
      sev = "error";
    else if(note.severity == Severity::warning)
      sev = "warning";
    result.items.push_back(make_item("compliance", note.message, false, "", sev));
  }

  // --- Naming checks ---
  if(opts.check_naming)
  {
    for(const auto& cpl_path : cpls)
    {
      auto cpl = Cpl::parse(cpl_path);
      if(cpl.has_value())
      {
        auto naming_notes = check_isdcf_naming(cpl->content_title, cpl_path);
        bool naming_ok = naming_notes.empty();
        result.items.push_back(make_item(
            "naming", "ISDCF naming compliance", naming_ok,
            naming_ok ? "" : std::to_string(naming_notes.size()) + " naming issue(s)", "warning"));
      }
    }
  }

  // --- Hash verification ---
  if(opts.check_hashes)
  {
    ChecksumVerifyOptions hash_opts;
    hash_opts.package_dir = opts.dcp_dir;
    auto hash_result = verify_package_checksums(hash_opts);
    result.items.push_back(
        make_item("integrity", "File hash verification", hash_result.all_valid,
                  hash_result.all_valid
                      ? ""
                      : std::to_string(hash_result.hash_mismatches) + " hash mismatch(es)"));
  }

  // --- Reel continuity ---
  for(const auto& cpl_path : cpls)
  {
    auto cont_notes = check_reel_continuity(cpl_path);
    result.items.push_back(make_item("continuity", "Reel continuity", cont_notes.empty(),
                                     cont_notes.empty() ? "" : "Reel continuity issues detected",
                                     "warning"));
  }

  // --- Summarize ---
  for(const auto& item : result.items)
  {
    result.checks_total++;
    if(item.passed)
    {
      result.checks_passed++;
    }
    else
    {
      if(item.severity == "error")
        result.errors++;
      else if(item.severity == "warning")
        result.warnings++;
      else
        result.info_count++;
    }
  }

  result.ready = (result.errors == 0);
  result.summary = std::to_string(result.checks_passed) + "/" +
                   std::to_string(result.checks_total) + " checks passed";
  if(result.errors > 0)
    result.summary += ", " + std::to_string(result.errors) + " error(s)";
  if(result.warnings > 0)
    result.summary += ", " + std::to_string(result.warnings) + " warning(s)";

  spdlog::info("Facility check: {} — {}", result.ready ? "READY" : "NOT READY", result.summary);
  return result;
}

std::string facility_check_to_json(const FacilityCheckResult& result)
{
  std::ostringstream json;
  json << "{\n";
  json << "  \"ready\": " << (result.ready ? "true" : "false") << ",\n";
  json << "  \"summary\": \"" << result.summary << "\",\n";
  json << "  \"errors\": " << result.errors << ",\n";
  json << "  \"warnings\": " << result.warnings << ",\n";
  json << "  \"checks_passed\": " << result.checks_passed << ",\n";
  json << "  \"checks_total\": " << result.checks_total << ",\n";
  json << "  \"items\": [\n";
  for(size_t i = 0; i < result.items.size(); ++i)
  {
    const auto& item = result.items[i];
    json << "    {\"category\": \"" << item.category << "\", "
         << "\"check\": \"" << item.check_name << "\", "
         << "\"passed\": " << (item.passed ? "true" : "false") << ", "
         << "\"severity\": \"" << item.severity << "\"";
    if(!item.detail.empty())
      json << ", \"detail\": \"" << item.detail << "\"";
    json << "}";
    if(i + 1 < result.items.size())
      json << ",";
    json << "\n";
  }
  json << "  ]\n";
  json << "}\n";
  return json.str();
}

} // namespace dcpdoctor

File: rust/crates/dcpdoctor-core/src/facility_check.rs
//! Pre-delivery facility check: comprehensive DCP validation for theater ingest.

use std::path::{Path, PathBuf};

use serde::Serialize;

/// A single check item result.
#[derive(Debug, Clone, Default, Serialize)]
pub struct CheckItem {
    pub category: String,
    pub check_name: String,
    pub passed: bool,
    pub detail: String,
    pub severity: String,
}

/// Options for facility check.
pub struct FacilityCheckOptions {
    pub dcp_dir: PathBuf,
    pub check_naming: bool,
    pub check_hashes: bool,
}

/// Result of facility check.
#[derive(Debug, Clone, Default, Serialize)]
pub struct FacilityCheckResult {
    pub error: String,
    pub ready: bool,
    pub summary: String,
    pub checks_total: u32,
    pub checks_passed: u32,
    pub errors: u32,
    pub warnings: u32,
    pub info_count: u32,
    pub items: Vec<CheckItem>,
}

fn make_item(category: &str, name: &str, passed: bool, detail: &str, severity: &str) -> CheckItem {
    CheckItem {
        category: category.into(),
        check_name: name.into(),
        passed,
        detail: detail.into(),
        severity: severity.into(),
    }
}

/// Run a comprehensive facility check on a DCP directory.
pub fn run_facility_check(opts: &FacilityCheckOptions) -> FacilityCheckResult {
    let mut result = FacilityCheckResult::default();

    if !opts.dcp_dir.exists() || !opts.dcp_dir.is_dir() {
        result.error = format!("DCP directory not found: {}", opts.dcp_dir.display());
        return result;
    }

    // --- Structure checks ---
    let has_assetmap =
        opts.dcp_dir.join("ASSETMAP").exists() || opts.dcp_dir.join("ASSETMAP.xml").exists();
    result.items.push(make_item(
        "structure",
        "ASSETMAP present",
        has_assetmap,
        if has_assetmap {
            ""
        } else {
            "Missing ASSETMAP or ASSETMAP.xml"
        },
        "error",
    ));

    let has_volindex =
        opts.dcp_dir.join("VOLINDEX").exists() || opts.dcp_dir.join("VOLINDEX.xml").exists();
    result.items.push(make_item(
        "structure",
        "VOLINDEX present",
        has_volindex,
        if has_volindex {
            ""
        } else {
            "Missing VOLINDEX or VOLINDEX.xml"
        },
        "error",
    ));

    // Check PKL
    let pkls = find_xml_rooted_at(&opts.dcp_dir, "PackingList");
    result.items.push(make_item(
        "structure",
        "PKL present",
        !pkls.is_empty(),
        if pkls.is_empty() {
            "No PackingList XML found"
        } else {
            ""
        },
        "error",
    ));

    // Check CPL
    let cpls = find_xml_rooted_at(&opts.dcp_dir, "CompositionPlaylist");
    result.items.push(make_item(
        "structure",
        "CPL present",
        !cpls.is_empty(),
        if cpls.is_empty() {
            "No CompositionPlaylist XML found"
        } else {
            ""
        },
        "error",
    ));

    // Check MXF files exist
    let has_mxf = std::fs::read_dir(&opts.dcp_dir)
        .into_iter()
        .flatten()
        .flatten()
        .any(|e| e.path().extension().is_some_and(|x| x == "mxf"));
    result.items.push(make_item(
        "structure",
        "MXF essence files present",
        has_mxf,
        if has_mxf {
            ""
        } else {
            "No .mxf files found in DCP directory"
        },
        "error",
    ));

    // --- Namespace consistency ---
    let ns_notes = crate::schema_validate::check_namespace_consistency(&opts.dcp_dir);
    let ns_ok = ns_notes.is_empty();
    result.items.push(make_item(
        "compliance",
        "Namespace consistency",
        ns_ok,
        if ns_ok {
            ""
        } else {
            "Mixed SMPTE/Interop namespaces"
        },
        "warning",
    ));

    // --- Hash verification ---
    if opts.check_hashes {
        let checksum = crate::checksum_verify::verify_package_checksums(
            &crate::checksum_verify::ChecksumVerifyOptions {
                package_dir: opts.dcp_dir.clone(),
                ..Default::default()
            },
        );
        let detail = if !checksum.success {
            format!("Hashes not verified: {}", checksum.error)
        } else if checksum.all_valid {
            String::new()
        } else {
            format!(
                "{} of {} asset(s) failed: {}",
                checksum.hash_mismatches + checksum.size_mismatches + checksum.missing_files,
                checksum.total_assets,
                failed_asset_details(&checksum).join(", ")
            )
        };
        result.items.push(make_item(
            "integrity",
            "PKL hash verification",
            checksum.success && checksum.all_valid,
            &detail,
            "error",
        ));
    } else {
        result.items.push(make_item(
            "integrity",
            "PKL hash verification",
            false,
            "Not checked (--no-hashes)",
            "info",
        ));
    }

    // --- Signature and signing certificates ---
    let mut signature_notes = Vec::new();
    let mut signed_documents = 0;
    for document in pkls.iter().chain(cpls.iter()) {
        let Ok(content) = std::fs::read_to_string(document) else {
            continue;
        };
        if !crate::signature::has_signature(&content) {
            continue;
        }
        signed_documents += 1;
        signature_notes.extend(crate::signature::verify_signature(document, false));
    }
    if signed_documents == 0 {
        result.items.push(make_item(
            "security",
            "Signing certificates",
            false,
            "Not checked: no CPL or PKL in the package carries a signature",
            "info",
        ));
    } else {
        let severity = if signature_notes
            .iter()
            .any(|note| note.severity == crate::Severity::Error)
        {
            "error"
        } else {
            "warning"
        };
        let detail = signature_notes
            .iter()
            .map(|note| note.message.clone())
            .collect::<Vec<_>>()
            .join(", ");
        result.items.push(make_item(
            "security",
            "Signing certificates",
            signature_notes.is_empty(),
            &detail,
            severity,
        ));
    }

    // --- ISDCF naming ---
    if opts.check_naming {
        for cpl_path in &cpls {
            let content = match std::fs::read_to_string(cpl_path) {
                Ok(c) => c,
                Err(e) => {
                    result.items.push(make_item(
                        "naming",
                        "ISDCF naming compliance",
                        false,
                        &format!("Not checked, cannot read {}: {e}", cpl_path.display()),
                        "warning",
                    ));
                    continue;
                }
            };
            let title_re =
                regex_lite::Regex::new(r"<ContentTitleText>([^<]+)</ContentTitleText>").unwrap();
            let Some(cap) = title_re.captures(&content) else {
                result.items.push(make_item(
                    "naming",
                    "ISDCF naming compliance",
                    false,
                    &format!("Not checked, no ContentTitleText in {}", cpl_path.display()),
                    "warning",
                ));
                continue;
            };
            let naming_notes = crate::isdcf::check_isdcf_naming(&cap[1], cpl_path);
            let naming_ok = naming_notes.is_empty();
            result.items.push(make_item(
                "naming",
                "ISDCF naming compliance",
                naming_ok,
                if naming_ok {
                    ""
                } else {
                    "ISDCF naming issues found"
                },
                "warning",
            ));
        }
    }

    // --- Summarize ---
    for item in &result.items {
        result.checks_total += 1;
        if item.passed {
            result.checks_passed += 1;
        } else {
            match item.severity.as_str() {
                "error" => result.errors += 1,
                "warning" => result.warnings += 1,
                _ => result.info_count += 1,
            }
        }
    }

    result.ready = result.errors == 0;
    result.summary = format!(
        "{}/{} checks passed",
        result.checks_passed, result.checks_total
    );
    if result.errors > 0 {
        result.summary += &format!(", {} error(s)", result.errors);
    }
    if result.warnings > 0 {
        result.summary += &format!(", {} warning(s)", result.warnings);
    }

    result
}

/// One line per asset that failed the checksum pass, naming the file and what
/// was wrong with it.
fn failed_asset_details(checksum: &crate::checksum_verify::ChecksumVerifyResult) -> Vec<String> {
    checksum
        .entries
        .iter()
        .filter_map(|entry| {
            let name = if entry.filename.is_empty() {
                entry.asset_id.clone()
            } else {
                entry.filename.clone()
            };
            if !entry.file_exists {
                return Some(format!("{name} is missing"));
            }
            if !entry.hash_match {
                return Some(format!("{name} hash mismatch"));
            }
            if !entry.size_match {
                return Some(format!(
                    "{name} is {} bytes, the PKL declares {}",
                    entry.actual_size, entry.expected_size
                ));
            }
            None
        })
        .collect()
}

/// Serialize facility check result to JSON.
pub fn facility_check_to_json(result: &FacilityCheckResult) -> String {
    serde_json::to_string_pretty(result).unwrap_or_default()
}

/// Package XML documents whose root element carries `root_name`. An ASSETMAP
/// holds a `<PackingList>` flag element of its own, so a substring search finds
/// it in place of a PKL that was never delivered.
fn find_xml_rooted_at(dir: &Path, root_name: &str) -> Vec<PathBuf> {
    let mut found = Vec::new();
    let Ok(entries) = std::fs::read_dir(dir) else {
        return found;
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if !path.is_file() || path.extension().is_none_or(|ext| ext != "xml") {
            continue;
        }
        if let Ok(content) = std::fs::read_to_string(&path)
            && crate::schema::root_element(&content).is_some_and(|(root, _)| root == root_name)
        {
            found.push(path);
        }
    }
    found.sort();
    found
}
